src/pykeen/nn/emb.py

```diff
@@ -1068,3 +1070,136 @@ def forward(
         if indices is not None:
             x = x[indices]
         return x
+
+
+class NodePieceRepresentation(RepresentationModule):
+    r"""
+    Basic implementation of node piece decomposition [galkin2021]_.
+
+    .. math ::
+        x_e = agg(\{T[t] \mid t \in tokens(e) \})
+
+    where $T$ are token representations, $tokens$ selects a fixed number of $k$ tokens for each entity, and $agg$ is
+    an aggregation function, which aggregates the individual token representations to a single entity representation.
+
+    .. note ::
+        This implementation currently only supports representation of entities by bag-of-relations.
+    """
+
+    #: the token representations
+    tokens: RepresentationModule
+
+    #: the entity-to-token mapping
+    assignment: torch.LongTensor
+
+    def __init__(
+        self,
+        *,
+        triples_factory: CoreTriplesFactory,
+        token_representation: Union[EmbeddingSpecification, RepresentationModule],
+        aggregation: Optional[Callable[[torch.Tensor, int], torch.Tensor]] = None,
+        num_tokens: int = 2,
+    ):
+        """
+        Initialize the representation.
+
+        :param triples_factory:
+            the triples factory
+        :param token_representation:
+            the token representation specification, or pre-instantiated representation module. For the latter, the
+            number of representations must be $2 * num_relations + 1$.
+        :param aggregation:
+            aggregation of multiple token representations to a single entity representation. By default,
+            this uses :func:`torch.mean`. It could also use other aggregations like :func:`torch.sum`,
+            :func:`torch.max`, or even trainable aggregations e.g., ``MLP(mean(MLP(tokens)))``
+            (cf. DeepSets from [zaheer2017]_) if given value ``"mlp"``.
+
+            We could also have aggregations which result in differently shapes output, e.g. a concatenation of all
+            token embeddings resulting in shape ``(num_tokens * d,)``. In this case, `shape` must be provided.
+
+            The aggregation takes two arguments: the (batched) tensor of token representations, in shape
+            ``(*, num_tokens, *dt)``, and the index along which to aggregate.
+        :param shape:
+            the shape of an individual representation. Only necessary, if aggregation results in a change of dimensions.
+        :param num_tokens:
+            the number of tokens for each entity.
+        """
+        mapped_triples = triples_factory.mapped_triples
+        if triples_factory.create_inverse_triples:
+            # inverse triples are created afterwards implicitly
```

Comment:
I would rather have the code explicitly disallow triples factories with create_inverse_triples rather than writing code to try and undo the operation

Reply:
This would prohibit using inverse triples for computing triple scores. The reference implementation always uses inverse trples, and also the implementation of the NodePiece uses `2 * R + 1` relation embeddings (with an additional padding relation embdding which is not part of any other model so far).